backend/src/api/matchCandidates.js

```diff
@@ -6,21 +6,83 @@ const { Candidate, Match } = require('../models')
 router.get(
   '/',
   errorWrap(async (req, res) => {
-    // not very efficient, nor should we have this algorithm for matches
-    // TODO: change this to a better algorithm
-    const candidates = await Candidate.aggregate([{ $sample: { size: 5 } }])
+    const min = await Candidate.aggregate([
+      {
+        $group: {
+          _id: {},
+          min: { $min: '$facemashRankings.numOfMatches' }
+        }
+      }
+    ])
+
+    const cand1 = await Candidate.aggregate([
+      { $match: { 'facemashRankings.numOfMatches': min[0].min } },
+      { $sample: { size: 1 } }
+    ])
+
+    const id1 = cand1[0]._id
+
+    const prev_matches = await Match.find({
+      $or: [{ candidate1: id1 }, { candidate2: id1 }]
+    })
+
+    const prev_ids = [id1]
+    for (let i = 0; i < prev_matches.length; i++) {
+      if (prev_matches[i].candidate1.equals(id1)) {
+        prev_ids.push(prev_matches[i].candidate2)
+      } else {
+        prev_ids.push(prev_matches[i].candidate1)
+      }
+    }
+
+    const secondMin = await Candidate.aggregate([
+      {
+        $match: {
+          _id: { $nin: prev_ids }
+        }
+      },
+      {
+        $group: {
+          _id: {},
+          min: { $min: '$facemashRankings.numOfMatches' }
+        }
+      }
+    ])
+
+    const cand2 = await Candidate.aggregate([
+      {
+        $match: {
+          $and: [{ 'facemashRankings.numOfMatches': secondMin[0].min }, { _id: { $nin: prev_ids } }]
+        }
+      },
+      { $sample: { size: 1 } }
+    ])
+
+    const id2 = cand2[0]._id
+
     // create the match
     const match = new Match({
-      candidate1: candidates[0]._id,
-      candidate2: candidates[1]._id
+      candidate1: id1,
+      candidate2: id2
     })
     await match.save()
+
+    const candidate1 = await Candidate.findById(id1)
```

Comment:
Can we use that? I wasn't sure because `aggregate` can do things like grouping and counting so I wasn't sure if we'd be able to modify and save the documents that `aggregate` returns.

Reply:
hmm I thought `aggregate` returns a `Candidate` that you can update. I might be wrong though.

Reply:
Okay, I'll look into it / try it out and see what happens.

Reply:
Tried it out, `aggregate` just returns the fields of the object, so I can't call `save` or anything else on the documents it returns.